source/slow-down.ts

```diff
@@ -0,0 +1,160 @@
+// /source/express-slow-down.ts
+// The speed limiting middleware.
+
+import type { Request, Response, NextFunction } from 'express'
+import { rateLimit, type Options as RateLimitOptions } from 'express-rate-limit'
+import type {
+	SlowDownRequestHandler,
+	AugmentedRequest,
+	SlowDownOptions,
+	Options,
+} from './types'
+
+/**
+ * Remove any options where their value is set to undefined. This avoids overwriting defaults
+ * in the case a user passes undefined instead of simply omitting the key.
+ *
+ * @param passedOptions {Options} - The options to omit.
+ *
+ * @returns {Options} - The same options, but with all undefined fields omitted.
+ *
+ * @private
+ */
+const omitUndefinedOptions = (
+	passedOptions: Partial<Options>,
+): Partial<Options> => {
+	const omittedOptions: Partial<Options> = {}
+
+	for (const k of Object.keys(passedOptions)) {
+		const key = k as keyof Options
+
+		if (passedOptions[key] !== undefined) {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+			omittedOptions[key] = passedOptions[key]
+		}
+	}
+
+	return omittedOptions
+}
+
+/**
+ * Create an instance of middleware that slows down responses to Express requests.
+ *
+ * @param passedOptions {Options} - Options to configure the speed limiter.
+ *
+ * @returns {SlowDownRequestHandle} - The middleware that speed-limits clients based on your configuration.
+ *
+ * @public
+ */
+export const slowDown = (
+	passedOptions: Partial<Options> = {},
+): SlowDownRequestHandler => {
+	// Passing undefined should be equivalent to not passing an option at all, so we'll
+	// omit all fields where their value is undefined.
+	const notUndefinedOptions: Partial<Options> =
+		omitUndefinedOptions(passedOptions)
+
+	if (
+		notUndefinedOptions.headers ||
+		notUndefinedOptions.legacyHeaders ||
+		notUndefinedOptions.standardHeaders
+	)
+		throw new Error(
+			'The headers options were removed in express-slow-down v2.0.0.',
+		)
+
+	if (
+		notUndefinedOptions.max !== undefined ||
+		notUndefinedOptions.limit !== undefined
+	)
+		throw new Error(
+			'The limit/max option is not supported by express-slow-down, please use delayAfter instead.',
+		)
+
+	// TODO: Remove in v3.
+	if (typeof notUndefinedOptions.delayMs === 'number') {
+		const url = `https://express-rate-limit.github.io/WRN_ESD_DELAYMS/`
+		console.warn(
+			`The behaviour of the 'delayMs' option was changed in express-slow-down v2. See ${url} for more information.`,
```

Comment:
Oh, right forgot about this. Thank you!

Two thoughts:

1) I think we should wrap it in `new Error(...)` to provide a stack trace.

2) Since this is going to hit *literally everyone*, I think we should provide the fix right in the error message:

```suggestion
			new Error(`The behaviour of the 'delayMs' option was changed in express-slow-down v2:

* For the old behavior, change the delayMs option to
  delayMs: (used, req) => {
	  const delayAfter = req.${notUndefinedOptions.requestPropertyName ?? 'slowDown'}.limit;
	  return (used - delayAfter) * ${notUndefinedOptions.delayMs};
  },

* For the new behavior, change the delayMs option to:
delayMs: () => ${notUndefinedOptions.delayMs},

See ${url} for more information.`),
```


3) It'd be nice if the warning could be disabled. Perhaps we can add a new `options.validate.delayMs` that we accept in ESD, but remove before passing along to ERL? And then maybe change the new behavior bullet point to point towards disabling the check rather than setting a function?

Reply:
Sorry for the late reply, I had an exam today and was preparing for it.

I changed the message to an error, so the stack trace shows up now. I also changed the message to what you suggested, and added the following at the end:

```
Set 'options.validate.delayMs: false' to disable this error message.
```

I added a way to disable the warning, and a test for it too.